microovn/ovn/certificates.go

```diff
@@ -0,0 +1,251 @@
+package ovn
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"database/sql"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"math/big"
+	"os"
+	"time"
+
+	"github.com/canonical/microcluster/state"
+	"github.com/canonical/microovn/microovn/database"
+	"github.com/canonical/microovn/microovn/ovn/paths"
+)
+
+const CACertName = "cacert"
+const CACertValidity = 10 * 365 * 24 * time.Hour     // 10 years
+const ServiceCertValidity = 2 * 365 * 24 * time.Hour // 2 years
+
+var MaxSerialNumber = new(big.Int).Lsh(big.NewInt(1), 128)
+
+type CertificateType int
+
+const (
+	CertificateTypeCA CertificateType = iota
+	CertificateTypeServer
+	CertificateTypeClient
+)
+
+// issueCertificate function generates new "server", "client" or "CA" certificate based on the value passed
+// to the "certType" argument. Argument "cn" is passed unchanged to the certificate's CN and "serviceName" is used in
+// certificate's OU.  Certificates are valid, based on their type, for a period specified in CACertValidity
+// and ServiceCertValidity.
+//
+// When generating certificate that is signed by a CA, "parent" argument must point to a parsed CA certificate and
+// "signer" argument must point to CA's private key. On the other hand if you want to generate self-signed certificate,
+// both "parent" and "signer" arguments must be empty (nil).
+//
+// This function returns PEM encoded certificate, private key and error (if any occurred).
+func issueCertificate(cn string, serviceName string, certType CertificateType, parent *x509.Certificate, signer *ecdsa.PrivateKey) ([]byte, []byte, error) {
+	var (
+		isCa     bool
+		keyUsage x509.KeyUsage
+		signKey  *ecdsa.PrivateKey
+		validTo  time.Time
+	)
+	// Generate certificate's private key
+	keyPair, err := ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to generate key pair for '%s' certificate: %w", serviceName, err)
+	}
+
+	// Generate certificate data and properties
+	validFrom := time.Now().UTC()
+	serialNumber, err := rand.Int(rand.Reader, MaxSerialNumber)
```

Comment:
The serial number must be unique per certificate per CA. As pointed out in the data structure above we should probably store this in the database and deduce the next available serial number from that information.

Reply:
In attempt to make the implementation stateless I took inspiration from [lxd](https://github.com/lxc/lxd/blob/123231d27121223f20f11a65499ce8a5e35d704b/shared/cert.go#L319). I think that with the max size of a serial being 2^128 it's quite safe to assume that we'll get unique serial numbers. On top of that I don't think that something actually breaks if there are two certificates with same serial number.

Reply:
:+1: this is fine. I remember talking to them and they brought up the requirements around serials, but if they don't care about it we probably don't need to for this version of the implementation :)